### ISSUE TITLE
lstsq sometimes overwrites its inputs

### ISSUE THREAD

**Issue by ah-** [numba/numba#3368]
Hi,

we've noticed that `np.linalg.lstsq` overwrites its input matrix in some edge cases. I'd expect the input matrix to never change.

To reproduce:

```python
import numpy as np
import numba as nb

X = np.array([[1., 7.54, 6.52],
              [1., 2.70, 4.00],
              [1., 2.50, 3.80],
              [1., 1.15, 5.64],
              [1., 4.22, 3.27],
              [1., 1.41, 5.70],
            ], order='F')

X_orig = X.copy()

y = np.array([1., 2., 3., 4., 5., 6.])

@nb.jit(nopython=True, nogil=True)
def f1(X, y):
    return np.linalg.lstsq(X, y)

f1(X, y)
(X == X_orig).all()
# True, all fine until here

@nb.jit(nopython=True, nogil=True)
def f2(X, y, test):
    if test:
        # never executed, but necessary to trigger the bug
        X = X[1:2, :]
    return np.linalg.lstsq(X, y)

f2(X, y, False)
(X == X_orig).all()
# False, X got corrupted
```

From the values in X it looks like it's likely being used as the work matrix of `lstsq`:
```python
array([[-2.44948974, 14.24766238,  0.53161088],
       [ 0.        ,  4.23875445,  2.46241227],
       [ 0.        ,  0.30113438,  3.52520967],
       [ 0.28989795,  0.47206594, -0.33747568],
       [ 0.28989795,  0.03894287,  0.52109546],
       [ 0.28989795,  0.43538451, -0.33387113]])
```

To trigger the issue it seems to be necessary for `X` to be in fortran order, and then be modified inside the jitted function.

I'm not entirely sure what causes this exactly, but looking at https://github.com/numba/numba/blob/master/numba/targets/linalg.py#L1619 it seems like it might be confused what the order of the X matrix is? And then skips doing the copy?

Tested with numba 0.39 as that's the latest available in anaconda, apologies if this is already fixed in 0.40.

**Comment by stuartarchibald** [comment]
Thanks for the report, I can confirm that this is indeed a bug (am on mainline builds).

I'm pretty sure that the `X = X[1:2, :]` on the dead branch causes the typing system to consider `X` as an `A` ordered array even though it is `F` ordered. Once that happens `np.asfortranarray` is called here, https://github.com/numba/numba/blob/master/numba/targets/linalg.py#L1623, which compiles down to a runtime check of the contiguousness of the array, which turns out to be `F` ordered already so it isn't copied, net result is `dgelsd` is called on the input array, hence mutation. I'll try and fix this for the 0.41 release and ping here when there's some dev builds with a fix in to check against.

In the mean time:
```python
@nb.jit(nopython=True, nogil=True)
def f2(X, y, test):
    if test:
        ## never executed, but necessary to trigger the bug
        X = X[1:2, :]
    return np.linalg.lstsq(np.asfortranarray(X), y)
```
should work around the problem as it'll make the typing system see `X` as `F` ordered and the linalg impl will pick this up and behave as expected.

**Comment by ah-** [comment]
Thanks for the super fast response! Happy to help test.
